pkg/querier/engine/streaming/operator/aggregation.go

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package operator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/util/zeropool"
+)
+
+type Aggregation struct {
+	Inner    InstantVectorOperator
+	Start    time.Time
+	End      time.Time
+	Interval time.Duration
+	Grouping []string
+
+	remainingInnerSeriesToGroup []*group // One entry per series produced by Inner, value is the group for that series
+	remainingGroups             []*group // One entry per group, in the order we want to return them
+}
+
+type group struct {
+	labels labels.Labels
+
+	// The number of input series that belong to this group that we haven't yet seen.
+	remainingSeriesCount uint
+
+	// Sum and presence for each step.
+	sums    []float64
+	present []bool
+}
+
+var _ InstantVectorOperator = &Aggregation{}
+
+var groupPool = zeropool.New(func() *group {
+	return &group{}
+})
+
+func (a *Aggregation) SeriesMetadata(ctx context.Context) ([]SeriesMetadata, error) {
+	// Fetch the source series
+	innerSeries, err := a.Inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer PutSeriesMetadataSlice(innerSeries)
+
+	if len(innerSeries) == 0 {
+		// No input series == no output series.
+		return nil, nil
+	}
+
+	// Determine the groups we'll return
+	groups := map[uint64]*group{}
+	buf := make([]byte, 0, 1024)
+	lb := labels.NewBuilder(labels.EmptyLabels())
+	a.remainingInnerSeriesToGroup = make([]*group, 0, len(innerSeries))
+
+	for _, series := range innerSeries {
+		var groupingKey uint64
+		groupingKey, buf = series.Labels.HashForLabels(buf, a.Grouping...)
```

Comment:
What's about hash collisions?

Reply:
As far as I can see, Prometheus' engine doesn't handle hash collisions either - so while that might be incorrect, both engines will at least be incorrect in the same way.

Reply:
Can you add a comment in the code to mention it? Like, we're aware of potential hash collisions here, but Prometheus engine does the same.

Reply:
Done: https://github.com/grafana/mimir/pull/7862